arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,234 @@
+
+      std::string::size_type shaIndex = it->find(".sha.");
+      if (shaIndex != std::string::npos) {
+        sstFileName = it->substr(0, shaIndex) + ".sst";
+        TRI_ASSERT(sstFileName == TRI_Basename(sstFileName));
+        auto nextIt = it + 1;
+        if (nextIt != fileList.end() && nextIt->compare(sstFileName) == 0) {
+          TRI_ASSERT(it->size() >= shaIndex + 64);
+          std::string hash = it->substr(shaIndex + /*.sha.*/ 5, 64);
+          it = nextIt;
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          _sstFileNamesToHashes.try_emplace(sstFileName, std::move(hash));
+        } else {
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("4eac9", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Deleting file "
+              << tempPath;
+          TRI_UnlinkFile(tempPath.c_str());
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          _sstFileNamesToHashes.erase(sstFileName);
+        }
+      } else if (isFileNameSst(*it)) {
+        auto hashIt = _sstFileNamesToHashes.find(*it);
+        if (hashIt == _sstFileNamesToHashes.end()) {
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("d6c86", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Computing cheacksum for "
+              << tempPath;
+          auto checksumCalc = ChecksumCalculator();
+          if (TRI_ProcessFile(tempPath.c_str(),
+                              [&checksumCalc](char const* buffer, size_t n) {
+                                checksumCalc.update(buffer, n);
+                                return true;
+                              })) {
+            writeShaFile(tempPath, checksumCalc.computeChecksum());
+          }
+        }
+      }
+    }
+  }
+}
+
+rocksdb::Status ChecksumEnv::NewWritableFile(
+    const std::string& fileName, std::unique_ptr<rocksdb::WritableFile>* result,
+    const rocksdb::EnvOptions& options) {
+  std::unique_ptr<rocksdb::WritableFile> writableFile;
+  rocksdb::Status s =
+      rocksdb::EnvWrapper::NewWritableFile(fileName, &writableFile, options);
+  if (!s.ok()) {
+    return s;
+  }
+  result->reset(
+      new ChecksumWritableFile(writableFile.release(), fileName, _helper));
+  return s;
+}
+
+rocksdb::Status ChecksumWritableFile::Close() {
+  if (!_helper->isFileNameSst(_sstFileName)) {
+    return rocksdb::WritableFileWrapper::Close();
+  }
+  TRI_ASSERT(_helper != nullptr);
+  auto checksumCalc = ChecksumCalculator();
+  if (TRI_ProcessFile(_sstFileName.c_str(),
+                      [&checksumCalc](char const* buffer, size_t n) {
+                        checksumCalc.update(buffer, n);
+                        return true;
+                      })) {
```

Comment:
Mmmmmh, this calculates the entire checksum upon closing the file only, which is inefficient.
Shouldn't we have already calculated the checksum incrementally when we get here?